### PR TITLE
Add argument include_tmax to .get_data methods

### LINKED ISSUE #10372: Adding an argument `include_tmax` for `.get_data()`
We ran into a failure today in https://github.com/vferat/pycrostates/pull/14 because `raw.get_data()` was returning one less sample than expected, namely, the `tmax` sample.

```
from pathlib import Path

import mne
print (mne.__version__)

directory = Path(mne.datasets.sample.data_path()) / 'MEG' / 'sample'
fname = directory / 'sample_audvis_filt-0-40_raw.fif' 
raw = mne.io.read_raw_fif(fname, preload=True, verbose=False)

print (raw.get_data().shape)
print (raw.get_data(tmin=0, tmax=raw.times[-1]).shape)
```

OUT:
```
0.24.0
(376, 41700)
(376, 41699)
```

Do you consider this to be a bug in the handling of `tmax`, or should an additional argument be added, `include_tmax`, like for `raw.crop()`?

### PR DESCRIPTION
**Pull request by mscheltienne**
Closes #10372.

Reworks `TimeMixin._handle_tmin_tmax` to support a `include_tmax` argument and to match the behavior described here:
https://github.com/mne-tools/mne-python/blob/79ea57a4318d8d045f5966c26360b079f40a4865/mne/utils/docs.py#L178-L183

`TimeMixin._handle_tmin_tmax` was only used in the `.get_data()` methods.
I've changed the function to use the method in `.crop()`. 

- [ ] Raw
- [x] Epochs
- [x] Evoked
- [ ] Changelog

### CODE REVIEW

File: mne/utils/docs.py [hunk 1]
```diff
@@ -167,17 +167,15 @@ def _reflow_param_docstring(docstring, has_first_line=True, width=75):
...
-    Python indexing typically works).
-
-    .. versionadded:: 0.19
+    If True (default), include ``tmax``. If False, exclude ``tmax`` (similar to
+    how Python indexing typically works).
 """
 docdict['notes_tmax_included_by_default'] = """
 Unlike Python slices, MNE time intervals by default include **both**
-their end points; ``crop(tmin, tmax)`` returns the interval
+their end points; ``inst.method(tmin, tmax)`` returns the interval
```
Comment (agramfort) [inline mne/utils/docs.py:178]: I find it less explicit for new comers what this inst.method means.

my 2c

Comment (mscheltienne) [reply mne/utils/docs.py:178]: True, I'll make a separate note for crop and get_data

**Comment by larsoner** [comment]
I have a feeling it will break a lot of people's code... probably worth a deprecation cycle `include_tmax=None` for it

**Comment by agramfort** [comment]
>
> I have a feeling it will break a lot of people's code... probably worth a
> deprecation cycle include_tmax=None for it
>

if you have the courage yes.... do we want to do this before 1.0?

**Comment by larsoner** [comment]
> if you have the courage yes.... do we want to do this before 1.0?

I think we need a deprecation cycle, too, since we had two "this is not a bug" comments in #10372

I think this would be a good PR for after the 1.0 release. That way there is no time pressure to deprecate properly

**Comment by mscheltienne** [comment]
Deprecation cycle is OK, I'll add it; I might not have the time to do it this week, so finishing this PR after 1.0 would be perfect.

**Comment by agramfort** [comment]
ok

**Comment by mscheltienne** [comment]
I've picked up this PR a bit yesterday, and there are different ways to add `include_tmax` for `.get_data()` methods. I started playing with `_time_mask`, and IMO, the implementation on `main` is flawed. Take the following example:

```
from mne.utils.numerics import _time_mask
import numpy as np


sfreq = 1
times = np.arange(0, 10 + 1 / sfreq, 1 / sfreq)

mask = _time_mask(times, tmin=1, tmax=3, sfreq=sfreq, include_tmax=True)
print (times[mask])  #  [1., 2., 3.]

mask = _time_mask(times, tmin=1.4, tmax=3, sfreq=sfreq, include_tmax=True)
print (times[mask])  #  [1., 2., 3.]

mask = _time_mask(times, tmin=1.51, tmax=3, sfreq=sfreq, include_tmax=True)
print (times[mask])  #  [2., 3.]

mask = _time_mask(times, tmin=1, tmax=3, sfreq=sfreq, include_tmax=False)
print (times[mask])  #  [1., 2.]

mask = _time_mask(times, tmin=1, tmax=3.4, sfreq=sfreq, include_tmax=False)
print (times[mask])  #  [1., 2.]

mask = _time_mask(times, tmin=1, tmax=3.51, sfreq=sfreq, include_tmax=False)
print (times[mask])  #  [1., 2., 3.]

mask = _time_mask(times, tmin=1, tmax=3.51, sfreq=sfreq, include_tmax=True)
print (times[mask])  #  [1., 2., 3., 4.]
```

IMO, `tmin` corresponds to the minimum time that should be returned (included) and `tmax` to the maximum time that should be returned (included or not depending on `include_tmax`). But this is not what `_time_mask` does.

I added a fix to this PR yesterday, but it does break a lot of tests as `_time_mask` is used left and right, and not only for `.crop()` or `.get_data()`. It also changes the behavior of many functions and is not backward compatible. Personally, I consider this to be a bugfix, and would skip any deprecation regarding `_time_mask`; if you do want to continue with this change. WDYT?

**Comment by larsoner** [comment]
> I added a fix to this PR yesterday, but it does break a lot of tests as _time_mask is used left and right, and not only for .crop() or .get_data(). It also changes the behavior of many functions and is not backward compatible

I'm worried about breaking things for people on a potentially large scale here. This is usually a sign that we have to be really sure that the change is worth it.

IIRC at some point we decided this `time_mask` behavior was reasonable -- if you want you could follow the `blame` and look into a bunch of PRs. But I think in the end it's probably not worth it.

So overall I'm -1 on changing `_time_mask` behavior, even if it's a bit unintuitive. I'd rather document somewhere how the public functions behave as a consequence of its behavior so that users can get the behaviors that they want/need. It seems like a safer, better use of dev (and user) time than risking breaking a bunch of stuff by changing how `_time_mask` works on such a large scale.

**Comment by mscheltienne** [comment]
So that's why I stopped after those 2 last commits, I wanted to gather opinions 😄 The change to `_time_mask` is not required to add `include_tmax` to `.get_data` methods anyway. If it's unnecessary, I'll revert the 2 last commits.

------------

So on one side, fixing `_time_mask` is not backward compatible, it will change the results from different functions. But is it actually going to break stuff? 
Sure, sample selection *might* differ by 1 sample. This 1 sample will result in (slightly) different outcomes for all the functions involved, but should not raise except in the case where we go from 1 sample returned to 0, an edge case which should not occur as long as users used *reasonable* `tmin` and `tmax` argument. 
I did not look in detail where `_time_mask` is used, do you foresee a scenario in which code would actually break?

Personally, I think it's a good quality of life improvement that makes the selection based on times easier. It removes the need for a note to all the functions that use time-based selection, a note that I fear will confuse more people than it will help.

**Comment by larsoner** [comment]
> So on one side, fixing _time_mask is not backward compatible, it will change the results from different functions. But is it actually going to break stuff?

I think:

1. It is very difficult to know/be convinced that we will *not* break stuff for people
2. The more a bit of code is used internally, the more likely that public interfaces will break in a way that is bad for people
3. If we attempt to change a behavior and some tests break, it's a sign that things are likely to break for people

Points (2) and (3) seem particularly worrisome here.

> Personally, I think it's a good quality of life improvement that makes the selection based on times easier. It removes the need for a note to all the functions that use time-based selection, a note that I fear will confuse more people than it will help.

I agree from a purity standpoint / in principle doing something different might have made sense if we were implementing from scratch. But given what we have, from a practical standpoint I don't think the (negligible?) gains from changing the implementation are worth the clear backward compat break. In other words, practicality beats purity for me here

**Comment by mscheltienne** [comment]
Alright, so let's revert those 2 last commits to keep `_time_mask` as is, and let's work towards a note that explains its behavior.
One more point, as `.get_data` method does use time-based selection, I changed its behavior by using `_time_mask`, at least this way MNE always uses the same function internally for time-based selection. 
This is part of the backward-incompatible change I originally introduced in this PR and that is to be covered by a deprecation cycle (discussed above). Are you still OK with this one?

File: mne/io/base.py [hunk 2]
```diff
@@ -92,28 +93,25 @@ def _handle_tmin_tmax(self, tmin, tmax):
...
-        n_times = self.times.size
-        start = 0 if tmin is None else self.time_as_index(tmin)[0]
-        stop = n_times if tmax is None else self.time_as_index(tmax)[0]
-
-        # truncate start/stop to the open interval [0, n_times]
-        start = min(max(0, start), n_times)
-        stop = min(max(0, stop), n_times)
+        start, stop = np.where(_time_mask(
+            self.times, tmin, tmax, sfreq=self.info['sfreq'],
+            include_tmax=include_tmax))[0][[0, -1]]
```
Comment (mscheltienne) [inline mne/io/base.py:114]: To highlight the code change, to use `_time_mask` internally for time-based selection.

**Comment by larsoner** [comment]
> Alternatively, I can get include_tmax argument in without this change as well.

I think there are two issues:

1. Adding `include_tmax`
2. Refactoring `get_data` to use `_time_mask`

To keep things as simple as possible, you could do (2) first. This should require no changes to any existing tests if the implementation is truly equivalent (I think/hopefully). Then we merge that PR, and then (1) presumably becomes more straightforward.

**Comment by mscheltienne** [comment]
Yes, I agree with those 2 steps. I did not have the time to pick it up, sorry. 
I actually wanted to go the other way around; revert the refactoring to `get_data` I made here and just add the `include_tmax`. I'm 99% sure the 2 implementations are not equivalent, and using `_time_mask` will fail some tests, again because of a diff of 1 sample.

**Comment by mmagnuski** [comment]
Hi, just adding my 2c to the discussion - I find the `_time_mask` examples reasonable, that's exactly what I would expect it to do (finding closest tmin / tmax time samples). I'm not sure in what sense the current behavior is flawed.

**Comment by mscheltienne** [comment]
Currently, `_time_mask` might return 1 sample before `tmin` and/or one sample after `tmax` depending on the time array and on `tmin`/`tmax`. It's really about edge cases, and the diff is only 1 sample anyway.. 
Also, I think (not tested) that some `crop` and some `get_data` methods might not have the same output for the same `tmin` / `tmax`, because they do not all use the same algo under the hood (e.g. `_time_mask`).

P.S: I will finish this PR at some point in the coming months 🙈

**Comment by mmagnuski** [comment]
I don't see the one sample before tmin and after tmax as a problem - if these samples are closer to tmin/tmax that's exactly what I would expect. :)
Lack of consistency between functions would be more problematic, however.

**Pull request closed without merging**